eclipse-theia/theia
# cli: introduce exclude id list for extension-packs

#### What it does


The pull-request introduces a mechanism to exclude extensions by `id` when resolving builtin **extension-packs**. The purpose is to allow application developers the flexibility to use extension-packs but still have the control to exclude problematic or unwanted extensions that may be referenced by the pack. Previously, post-processing would be necessary to remove such extensions, for example by deleting the plugin in the plugins directory before a build.

The changes introduce `theiaPluginsExcludeIds` which lists extensions by `id` and is used to exclude extensions when resolving extension-packs. 

#### How to test


- update the [`package.json`](https://github.com/eclipse-theia/theia/blob/471f10ae5beb24f7b072e80acf0e8b22c31e58ba/package.json#L87) to the following:

```json
"theiaPlugins": {
  "pack": "https://github.com/vince-fugnitto/vscode-builtin-extensionPack/releases/download/0.0.1/builtin-extension-pack-1.45.1.vsix"
},
"theiaPluginsExcludeIds": [
  "vscode.cpp"
]
```

- execute the cli command `yarn download:plugins`
- confirm that the pack is downloaded, and individual extensions are resolved
- confirm that `vscode.cpp` is logged as skipped
- confirm that `vscode.cpp` is not present under the `plugins` folder

_example:_

![image](https://user-images.githubusercontent.com/40359487/130655941-af0a0413-3556-4247-84fd-8c4cf7ca2ce1.png)


#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>
